File: leetcode/validate-binary-search-tree.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def inOrderTraversal(root, res):
            if not root:
                return

            inOrderTraversal(root.left, res)
            res.append(root.val)
            inOrderTraversal(root.right, res)

        elements = []
        inOrderTraversal(root, elements)

        length = len(elements)

        for i in range(1, length):
            if elements[i] <= elements[i - 1]:
                return False

        return True

File: leetcode/validate-binary-search-tree.cpp
#include <bits/stdc++.h>

using namespace std;

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
 public:
    bool isValidBST(TreeNode *root)
    {
        vector<int> res;
        int length;

        function<void(TreeNode *)> inorder = [&](TreeNode *root) {
            if (!root)
                return;

            inorder(root->left);
            res.push_back(root->val);
            inorder(root->right);
        };

        inorder(root);

        length = res.size();

        for (auto i = 1; i < length; ++i) {
            if (res[i] <= res[i - 1])
                return false;
        }

        return true;
    }
};
